### PR TITLE
Asynchronous LogStreamWriter#tryWrite

### LINKED ISSUE #10923: `LogStreamWriter#tryWrite` should return an actor future
Since we are planning to implement `Sequencer` as an actor, we want to change the interface of `LogStreamWriter` so that `tryWrite` returns a future. As long as we use the dispatcher, this will simply return an already completed future.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR modifies the `LogStreamWriter` to make it an asynchronous writer, in preparation for asynchronous writes with the dispatcher replacement. To achieve this:

- We had to remove usage of the `RetryStrategy` since they don't work with asynchronous operations. Instead, each caller must implement retry themselves by using `actor.runOnCompletion`, and in the callback, checking if the exception is recoverable, and if so calling `actor.run(myCallback); actor.yieldThread();`
- Introduce a flag in the scheduling service to avoid overlaps between retries (piggybacking off of the stream processor phase logic)
- Introduce actors/concurrency control to the inter partition command sender implementation

## Related issues

closes #10923 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
I'm not very happy with the outcome, to be honest. It really highlights the weaknesses of the actor scheduler as we built it.

- Cannot chain/transform futures makes for very verbose code
- Blurs the separation between synchronous and asynchronous code in the inter-partition command sender (bad for maintainability and test-ability)
- Would require more work to interoperate with the `RetryStrategy`, meaning everywhere you need a retry, you need to "re-implement" it yourself (and be careful about overlaps between operations)
- The API itself is currently very silly, using exceptions for flow control (a big no-no). I think we thought this would be different later on, but now I'm actually not sure.

File: logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamBatchWriterImpl.java [hunk 1]
```diff
@@ -200,31 +201,35 @@ private void copyExistingEventToBuffer() {
   }
 
   @Override
-  public long tryWrite() {
+  public ActorFuture<Long> tryWrite() {
     if (eventCount == 0) {
       if (valueWriter == null) {
-        return 0;
+        return CompletableActorFuture.completed(0L);
```
Comment (npepinpe) [inline logstreams/src/main/java/io/camunda/zeebe/logstreams/impl/log/LogStreamBatchWriterImpl.java:207]: This API makes no sense :smile:

**Comment by npepinpe** [comment]
Honestly, I wonder if we shouldn't drop this and just keep the same behavior as the dispatcher: synchronous writes, asynchronous reads. :shrug: 

Since we don't know what will be our flow control, and we don't know what we will have as a concurrency model in the end, I'm not sure this PR is useful at the moment.

**Comment by Zelldon** [comment]
> Honestly, I wonder if we shouldn't drop this and just keep the same behavior as the dispatcher: synchronous writes, asynchronous reads. :shrug: 
> 
> Since we don't know what will be our flow control, and we don't know what we will have as a concurrency model in the end, I'm not sure this PR is useful at the moment.

This is what I thought when raising this topic. Just having a queue with record batch objs

**Comment by npepinpe** [comment]
What would flow control look like with option 4? Option 2 seems fairly obvious, we could also have flow control return a specific value instead of throwing (and throw only for exceptional cases).

So for option 4, how would you avoid overwhelming the writer with requests? It's not really an issue for everywhere we retry, since I think there we always wait for the write to complete before enqueuing another one, but it could be for the request APIs or the inter partition command sender.

- Stream processor: we always write and retry until successful or an unrecoverable error occur, so we never enqueue more than one write at a time.
- Command API: on exceptional failure we return an error, otherwise, even if we did not write the command to the dispatcher, we return -1. I suppose this works because the request gets retried eventually, or simply times out. :shrug: 
- Backup API: same as above.
- Inter partition command sender: all requests are fully async, in that they don't expect a reply. So on exceptional error we log an error, and everything else we just log a warning.

For the APIs and the inter-partition case, we could technically overload the new dispatcher since we're not waiting for the write to complete before enqueuing the next one. This means growing the sequencer's actor queue, I guess? 

It's also a bit difficult to do with the current implementations, because the writers are not actor-aware, so how are they meant to loop/retry? Their callers must do so. I could pass adapters, I suppose, which are actor aware, but that seems like an awful lot of refactoring.

I'm thinking this may be us getting ahead of ourselves as we still don't know what the flow control mechanism will even look like, so we can't predict the best solution?

@Zelldon - regarding your point, one thing that stood out to us when prototyping is, why is the dispatcher asynchronous for reads (via actors) but synchronous for writes? Why use one concurrency model in one case, and not in the other? Since we have to update more than one field atomically when writing, and we already have to interface with actors, it seemed natural to have writes also be synchronized via the actor scheduler. But it does make consuming the writer API much, much more complicated (no surprise there).

**Comment by oleschoenburg** [comment]
> So for option 4, how would you avoid overwhelming the writer with requests?

Yeah, I guess that is the crux :sweat_smile: It's also what I meant by:
> and even for the coming sequencer, it's not ideal because the sequencer would have to buffer pending writes which sort of defeats one of the purposes of adding flow control.

I'm okay with leaving the `tryWrite` as is for now. Ideally we could end up with option 2 and optionally add a non-blocking version (think `offer` vs `put`) if that's really necessary.

Sorry you put so much work into this already, it wasn't clear to me how tricky it'd be to adapt the existing writers.

**Comment by Zelldon** [comment]
> @Zelldon - regarding your point, one thing that stood out to us when prototyping is, why is the dispatcher asynchronous for reads (via actors) but synchronous for writes? Why use one concurrency model in one case, and not in the other? Since we have to update more than one field atomically when writing, and we already have to interface with actors, it seemed natural to have writes also be synchronized via the actor scheduler. But it does make consuming the writer API much, much more complicated (no surprise there).

I guess I wouldn't design it that way. For me it is just a ManyToOneQueue, concurrency data structure where you can have multiple writers but only one reader. 

I think right now the dispatcher is also similar to that, the LogAppender is an actor that is true but it is just the consumer the reading (or polling) from the dispatcher (or subscription) is synchronous.

**Comment by npepinpe** [comment]
> I think right now the dispatcher is also similar to that, the LogAppender is an actor that is true but it is just the consumer the reading (or polling) from the dispatcher (or subscription) is synchronous.

The dispatcher is also an actor though; it always seemed strange to me that it's an actor, but writes bypass that entirely. I wanted to try and see if we used purely the actor as synchronization mechanism how it would be. From this PR, it's not...great. But at the same time, the worst of it comes down to only a few things:

- It's being used in contexts that were not actor aware previously (so, I guess more refactoring is to be expected there)
- Our actor scheduler API is awkward (e.g. cannot chain operations, have to be careful to use from an actor context/outside an actor context, etc.)
- We use exceptions as a means of flow control (e.g. with `Either`, with futures, etc.), and not just for exceptional cases. This isn't unique to this PR though.

> Sorry you put so much work into this already, it wasn't clear to me how tricky it'd be to adapt the existing writers.

@oleschoenburg - don't worry, I only put a few hours in, it's not that much! I think we both wanted to try and get a feel for how it would be to use the actor scheduler end-to-end, so I don't think this is lost effort either :+1:

**Comment by npepinpe** [comment]
I'll close the PR and the issue for now, but keep the branch a while longer, at least until we're done with the epic. We can revisit it once we know what the flow control mechanism will look like.

**Comment by npepinpe** [comment]
A good example of issues which arise due to mixing both concurrency models in one: https://github.com/camunda/zeebe/issues/11091

**Pull request closed without merging**